modules/renter/contractor/downloader.go

```diff
@@ -51,25 +58,63 @@ func (hd *hostDownloader) Sector(root crypto.Hash) ([]byte, error) {
 
 // Close cleanly terminates the download loop with the host and closes the
 // connection.
-func (hd *hostDownloader) Close() error { return hd.downloader.Close() }
+func (hd *hostDownloader) Close() error {
+	hd.mu.Lock()
+	defer hd.mu.Unlock()
+	// Close is a no-op unless there is only one goroutine using the
+	// hostDownloader
+	hd.clients--
+	if hd.clients > 0 {
+		return nil
+	}
+	hd.contractor.mu.Lock()
+	delete(hd.contractor.downloaders, hd.contractID)
+	delete(hd.contractor.revising, hd.contractID)
+	hd.contractor.mu.Unlock()
+	return hd.downloader.Close()
+}
 
-// Downloader initiates the download request loop with a host, and returns a
-// Downloader.
-func (c *Contractor) Downloader(contract modules.RenterContract) (Downloader, error) {
+// Downloader returns a Downloader object that can be used to download sectors
+// from a host.
+func (c *Contractor) Downloader(id types.FileContractID) (Downloader, error) {
 	c.mu.RLock()
+	cachedDownloader, haveDownloader := c.downloaders[id]
 	height := c.blockHeight
+	contract, haveContract := c.contracts[id]
 	c.mu.RUnlock()
-	if height > contract.EndHeight() {
+	host, haveHost := c.hdb.Host(contract.NetAddress)
+
+	if haveDownloader {
+		// increment number of clients
+		cachedDownloader.mu.Lock()
+		cachedDownloader.clients++
+		cachedDownloader.mu.Unlock()
+		return cachedDownloader, nil
+	} else if !haveContract {
+		return nil, errors.New("no record of that contract")
+	} else if height > contract.EndHeight() {
 		return nil, errors.New("contract has already ended")
-	}
-	host, ok := c.hdb.Host(contract.NetAddress)
-	if !ok {
+	} else if !haveHost {
 		return nil, errors.New("no record of that host")
-	}
-	if host.DownloadBandwidthPrice.Cmp(maxDownloadPrice) > 0 {
+	} else if host.DownloadBandwidthPrice.Cmp(maxDownloadPrice) > 0 {
 		return nil, errTooExpensive
 	}
 
+	// acquire revising lock
+	c.mu.Lock()
+	alreadyRevising := c.revising[contract.ID]
+	if alreadyRevising {
+		c.mu.Unlock()
+		return nil, errors.New("already revising that contract")
```

Comment:
hmm. So if a download is already in progress, you still get an error returned for this contract?

Seems like it should block until the revising resource becomes available. And if that's going to be in place, clean-shutdown needs to be considered, since you don't want to wait for all of your open downloads to complete before things close all the way.